prov/efa/test/efa_unit_test_common.c

```diff
@@ -1,10 +1,55 @@
 #include "efa_unit_tests.h"
 
-int efa_unit_test_resource_construct(struct efa_resource *resource)
+static
+struct fi_info *efa_unit_test_alloc_hints(enum fi_ep_type ep_type)
+{
+	struct fi_info *hints;
+
+	hints = calloc(sizeof(struct fi_info), 1);
+	if (!hints)
+		return NULL;
+
+	hints->domain_attr = calloc(sizeof(struct fi_domain_attr), 1);
+	if (!hints->domain_attr) {
+		fi_freeinfo(hints);
+		return NULL;
+	}
+
+	hints->fabric_attr = calloc(sizeof(struct fi_fabric_attr), 1);
+	if (!hints->fabric_attr) {
+		fi_freeinfo(hints);
+		return NULL;
+	}
+
+	hints->ep_attr = calloc(sizeof(struct fi_ep_attr), 1);
+	if (!hints->ep_attr) {
+		fi_freeinfo(hints);
+		return NULL;
+	}
+
+	hints->fabric_attr->prov_name = EFA_PROV_NAME;
+	hints->ep_attr->type = ep_type;
+
+	if (ep_type == FI_EP_DGRAM) {
+		hints->domain_attr->mr_mode |= FI_MR_LOCAL | FI_MR_ALLOCATED;
+		hints->mode |= FI_MSG_PREFIX;
+	} else {
+		hints->domain_attr->mr_mode |= FI_MR_LOCAL | FI_MR_ALLOCATED;
+	}
+
+	return hints;
+}
+
+int efa_unit_test_resource_construct(struct efa_resource *resource, enum fi_ep_type ep_type)
```

Comment:
Using `(struct efa_resource *resource, enum fi_ep_type ep_type)` as input prevents the user from configuring the resources using a different hint, e.g. passing `NULL`.

For now we can ask the user to call the setup funcs separately, i.e.
```c
...
hints = efa_unit_test_alloc_hints(ep_type);
ret = efa_unit_test_resource_construct(&resource, &hints) // or the user just needs efa_unit_test_resource_construct(&resource, NULL);
...
```

Reply:
I'm all for making user's life easier. But I prefer to do it differently. It's better to decouple resource construct from generating hint, and provide a shortcut function to the user - she can choose to use the shortcut, or do other customization.

For example, we can have 4 functions
```c
efa_unit_test_alloc_hints // same as now
efa_unit_test_resource_construct(struct efa_resource *resource, struct fi_info *hints) // Same as before, takes resource and hint
efa_unit_test_resource_enable // Binds resources and enables the endpoint
efa_unit_test_resource_setup(struct efa_resource *resource, enum fi_ep_type ep_type) // A shortcut to create hint, construct resources, bind and enable resources.
```

Reply:
> It's better to decouple resource construct from generating hint

Why is it better?

What customization is needed?

The proposed solution introduce 3 functions: construct, enable, setup. I think user will be really confused about which one to use.

Reply:
Better in the sense of having a single responsibility for a given function, e.g. `*_construct` only does resource creation, `*_enable` does what the name suggests - enabling the resources, etc. Provided that we document the code properly, the user should be able to understand with minimal effort.